### PR TITLE
Squash optional query params

### PR DESCRIPTION
**Pull request by sakari**
allow no query parameters when all query parameters are optional
see https://github.com/smartlyio/oats/issues/147

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -3,8 +3,16 @@ import * as assert from 'assert';
 import safe from '@smartlyio/safe-navigation';
 
 type HeaderProp<H, Next> = H extends void ? Next : { headers: H } & Next;
-type QueryProp<Q, Next> = Q extends void ? Next : { query: Q } & Next;
-type BodyProp<B> = B extends void ? object : { body: B };
+type QueryProp<Q, Next> = true extends HasOnlyOptionalTypes<Q>
```
Comment (sakari) [inline src/client.ts:6]: basically try to convince the ts compiler to allow `.get()`, `.get({query: nn})`, and `.get({query:nn, otherstuff })` when all the query parameters are optional

File: src/client.ts [hunk 2]
```diff
@@ -3,8 +3,16 @@ import * as assert from 'assert';
...
+  ? { query?: Q } | ({ query?: Q } & Next) | Next
+  : { query: Q } | ({ query: Q } & Next);
+type BodyProp<B> = B extends void ? void : { body: B };
+
+type HasOnlyOptionalTypes<O> = {
+  [K in keyof O]?: O[K];
+} extends O
+  ? true
+  : false;
```
Comment (ganymedes01) [inline src/client.ts:15]: Could probably simplify this to
```suggestion
type HasOnlyOptionalTypes<O> = Partial<O> extends O ? true : false;
```

File: src/client.ts (hunk 1 above)
Comment (ganymedes01) [reply src/client.ts:6]: Those types got really nested and hard to follow. Do you think something like this is more straight forward to read? It should (hopefully) behave the same

```ts
type NotVoidKeys<T> = {
  [P in keyof T]: T[P] extends void ? never : P;
}[keyof T];
type RemoveVoidProps<T> = Pick<T, NotVoidKeys<T>>;

type HasOnlyOptionalTypes<O> = Partial<O> extends O ? true : false;

export type ClientArg<
  H extends server.Headers | void,
  Q extends server.Query | void,
  B extends server.RequestBody<any> | void
> = RemoveVoidProps<
  {
    headers: H;
    body: B;
  } & (true extends HasOnlyOptionalTypes<Q> ? { query?: Q } : { query: Q })
>;
```

**Review by ganymedes01 — APPROVED** [review]
LGTM 💯
I'd "simplify" the types, but not really that fussed about it.

**Review by khugo — APPROVED** [review]
Seems to fulfill my needs, but I'll leave the code review to the experts :+1:
